doc/agent.rst

```diff
@@ -249,104 +259,55 @@ and copies them into ``Message.Parameters``. It then sets ``Message.Class`` to
-terminate. The scheduler sends a kill action to both agents with the PID of the
-old agent (8). The kill action may be executed twice, but that doesn't matter.
-When the scheduler receives the kill results (9), it sends a new action to check
-for `mig-agent` processes (10). Only one should be found in the results (11),
-and if that is the case, the scheduler tells the agent to remove the binary of
-the old agent (12). When the agent returns (13), the upgrade protocol is done.
-
-If the PID of the old agent lingers on the system, an error is logged for the
-investigator to decide what to do next. The scheduler does not attempt to clean
-up the situation.
+Agent upgrade process via mig-loader
+------------------------------------
+MIG supports upgrading agents in the wild through the use of the companion
+program mig-loader. Using mig-loader is optional; you don't need to use
+mig-loader in your environment if you want to upgrade agents yourself.
+
+The following is a high level diagram of how the loader interacts with the
+API and the agent during the upgrade process.
+
+::
+
+        Agent                Loader              API
+        +---+                +----+             +--+
+        |                    |                     |
+        |                    | 1. request manifest |
+        |                    |-------------------->|------+
+        |                    |                     |      | 2. update loader
+        | 3. valid  +--------|                     |      | record in database
+        | manifest  |        |                     |<-----+
+        | sig?      +------->|                     |
+        |                    |                     |
+        | 4. does   +--------|                     |
+        | current   |        |                     |
+        | agent     |        |                     |
+        | match?    +------->|                     |
+        |                    |                     |
+        |                    | 5. fetch new agent  |
+        |                    |-------------------->|
+        |                    |                     |
+        | 6. stage  +--------|                     |
+        | agent on  |        |                     |
+        | disk      +------->|                     |
+        |                    |                     |
+        | 7. agent  +--------|                     |
+        | SHA256    |        |                     |
+        | matches   |        |                     |
+        | manifest? +------->|                     |
+        |                    |                     |
+        |  8. install agent  |                     |
+        |<-------------------|                     |
+        |                    |                     |
+        |  9. stop old agent |                     |
+        |<-------------------|                     |
+        |                    |                     |
+        | 10. start new      |                     |
```

Comment:
Start may be the wrong term in the doc here; what actually happens is the loader starts the agent, and the agent installs itself as a service to run and exits. In normal circumstances, after this the service mgr will start the agent up (systemd, upstart, etc).

If for some reason this fails, the intent is since the loader is in cron/systemd timer it would run again. When that happens it will detect the agent isn't running and repeat the process.

This would correct any intermittent sort of error, if it's something going wrong we can't recover it unless we deploy new loader/agent code. This happens before the start, so if the error is resolved we can fix it and it would correct on next run after the manifest is updated.